### ISSUE TITLE
ARTEFACTS when playing: the plane changes position very quickly or "jumps"

### ISSUE THREAD

**Issue by Raygou** [nguyenquyhy/Flight-Recorder#32]
I am Raymond66 on https://flightsim.to/ 
This is about the ARTEFACTS when playing: the plane changes position very quickly or "jumps".
Could this be due to missing data?
I attach the 2 files of the same recording you asked me: in a zip file
[Extra 330LT Courchevel Export 2021-02-24-21-23.zip](https://github.com/nguyenquyhy/Flight-Recorder/files/6049679/Extra.330LT.Courchevel.Export.2021-02-24-21-23.zip)

I read you extrapolate data, so I assume there ARE data missing, so you have to have an algorithm to guess the missing data.
Could the artefacts be caused by a - instead of a +, and vice versa?

I precise I use v 0.9.   Let me know; I can test a new version with this type of flying...

**Comment by nguyenquyhy** [comment]
![main_full](https://user-images.githubusercontent.com/1560418/109396519-01ace100-7900-11eb-9031-fc289bd4d095.gif)

Better?

**Comment by Raygou** [comment]
OH dear, what is going on in your post above?

I have done another recording:
https://www.youtube.com/watch?v=N_tSq2hw6XM

with the 2 files .csv and .flightrecorder, in a zip:
[2 Extra 330LT Courchevel Export 2021-02-26-20-26.zip](https://github.com/nguyenquyhy/Flight-Recorder/files/6055149/2.Extra.330LT.Courchevel.Export.2021-02-26-20-26.zip)

Did you find what is happening?
It would be perfect without these artefacts.

**Comment by nguyenquyhy** [comment]
The gif is just a funny bug due to some typo in my calculation 🤣. 

And the artifact is fixed. Actually I know it would happen when I wrote the interpolation code, but just wanted to share the tool early and wasn't expect so many would find it useful, so I didn't want to fix it yet 😂. I knew what to fix immediately once I see your video.

https://youtu.be/kwM2IZrTNG4

I'll push the code later today and will release 0.10.0 tomorrow once I added some more features/fixes.

**Comment by Raygou** [comment]
Yeah I tough the was a funny thing. I did wonder if planes started to drink...You never know these days!

Ah this is great news. What is the percentage of missing data?
Take your time; rushing is never a good idea. I will test it when it will be released.
**I just wonder if it's possible to stop the alarm sounding**, because from outside there is no alarms in real life.
**Your tool is beyond extremely useful!**

**Comment by nguyenquyhy** [comment]
About the alarm, are you saying that when you are recording, there is no alarm? I think the sound is the stall warning, right?

**Comment by Raygou** [comment]
Sorry, I just see your question now... Yes, the sound is the stall warning.
The stall warning is recorded, so, when replaying:
from the cockpit view, the stall warning is wanted, but from outside view, the stall warning is not wanted...
What could you do?

**Comment by nguyenquyhy** [comment]
You can just turn off "Warning sound in external view" in your sound setting 😄.

**Comment by Raygou** [comment]
Actually, yes. Sound settings is the part I never really looked at.

**Comment by nguyenquyhy** [comment]
Released in v0.10

**Comment by Raygou** [comment]
OK, I will try v0.10... Now, to save time: Are the artefacts in the Replay, or in the saved .flightrecorder files?
If I know this I could avoid doing a new recording, and just replay the recording I have.
Of course if the artefacts come from the data, I will have to do a new recording...

**Comment by nguyenquyhy** [comment]
It's purely a problem with replay. All the recording is fine.

**Comment by Raygou** [comment]
Thank you so much, and for reply that quickly. That's what I call Diamond Service. I just have to do something else, then I try...

**Comment by Raygou** [comment]
I see my recording has an average gap  of 38.5 milliseconds; that gives ~ 25.97 FPS on average (my graphics settings are high to ultra).
**I think you record the data at each FPS? Is this correct?**
If so, then it may help to have smoother replays by turning down the graphics setting when actually recording the data when flying in FS, then increase graphics settings when replaying. This would keep the high FPS of the recording, and replaying in better graphics quality.
Am I correct, or would FS slow down, even when just replaying, in high graphics settings?

**Comment by nguyenquyhy** [comment]
The recorded data is based on sim frame, which involves simulation calculation, and is around 30fps and usually different from "graphic" fps. 
Turning down the graphic wouldn't affect much unless your CPU is bottleneck heavily cannot handle 30fps simframe.

For replaying, I am syncing with a Frame event, which I am not sure if that is sim frame or graphic frame. But regardless, the frame timing between recording and replaying never match, so the tool uses interpolation to guesstimate the value at the replay frame.

In other words, the replay will send data at the same rate your sim can handle during the replay. That rate doesn't change even if the recording rate is much higher or lower.

**Comment by Raygou** [comment]
OK, interesting. Then I think reducing graphics settings won't help much. That is why I ask, rather than doing... days of testing... for nothing, or not much, to gain. My CPU is ~ 5 years old: i7 5960X; it was one of the best at the time.
Is it a "linear" interpolation you do, or a more complex one, looking at several steps back to know if a parameter is stable, accelerating, or decelerating?

**Comment by Raygou** [comment]
See the brilliant result:
https://www.youtube.com/watch?v=OSUV0k28hrA
